include/seqan3/io/stream/iterator.hpp

```diff
@@ -166,6 +169,138 @@ class fast_istreambuf_iterator
+    fast_ostreambuf_iterator & operator=(fast_ostreambuf_iterator const &) noexcept = default; //!< Defaulted.
+    fast_ostreambuf_iterator & operator=(fast_ostreambuf_iterator &&) noexcept = default; //!< Defaulted.
+    ~fast_ostreambuf_iterator() noexcept = default; //!< Defaulted.
+
+    //!\brief Construct from a stream buffer.
+    explicit fast_ostreambuf_iterator(std::basic_streambuf<char_t, traits_t> & ibuf) :
+        stream_buf{reinterpret_cast<stream_buffer_exposer<char_t, traits_t> *>(&ibuf)}
+    {
+        assert(stream_buf != nullptr);
+        if (stream_buf->pptr() == stream_buf->epptr())
+            stream_buf->overflow(); // ensures that put area has space available
+    }
+    //!\}
+
+    /*!\name Arithmetic operators
+     * \{
+     */
+    //!\brief no op.
+    fast_ostreambuf_iterator & operator++()
+    {
+        return *this;
+    }
+    //!\overload
+    fast_ostreambuf_iterator & operator++(int)
+    {
+        return *this;
+    }
+    //!\}
+
+    //!\brief no op.
+    fast_ostreambuf_iterator & operator*()
+    {
+        return *this;
+    }
+
+    //!\brief Writes a character to the associated output stream.
+    fast_ostreambuf_iterator & operator=(char_t const c)
+    {
+        assert(stream_buf != nullptr);
+        if (stream_buf->pptr() == stream_buf->epptr())
+        {
+            stream_buf->sputc(c); // overflow() [virtual], then write character
+        }
+        else
+        {
+            *stream_buf->pptr() = c;
+            stream_buf->pbump(1); // advance pptr() in put area without any checks
+        }
+        return *this;
+    }
+
+    //!\brief Returns `true if this iterator has encountered the end-of-file condition on output, `false` otherwise.
+    bool failed() const noexcept
+    {
+        return stream_buf->overflow() == traits_t::eof();
+    }
+
+    /*!\brief Writes exactly `n` characters `[it, it + n)` to the associated output.
+     * \tparam iterator_type The iterator type; Must model std::ranges::input_iterator.
+     * \param[in, out] it The iterator to the range to write.
+     */
+    template <std::ranges::input_iterator iterator_type>
+    void write_n(iterator_type & it, size_t n)
+    {
+        if (size_t buffer_space = stream_buf->epptr() - stream_buf->pptr(); n <= buffer_space)
+        {
+            it = std::ranges::copy_n(it, n, stream_buf->pptr()).in; // copy_n returns incremented iterators {in, out}
+            stream_buf->pbump(n);
+        }
+        else
+        {
+            it = std::ranges::copy_n(it, buffer_space, stream_buf->pptr()).in;
+            stream_buf->pbump(buffer_space);
+
+            if (stream_buf->overflow(*it) == traits_t::eof()) // update put area by trying to write the next char
+                throw std::ios_base::failure{"Could not write to output stream."};
+
+            ++it;
+            ++buffer_space;
+
+            write_n(it, n - buffer_space);
```

Comment:
```suggestion
            /* Write remaining input data recursively */
            write_n(it, n - buffer_space);
```

Reply:
Some documentation would help understanding the semantics here!

Reply:
You removed the documentation again in the latest version :roll_eyes:

Reply:
well because I removed write_n?